### PR TITLE
chore: create `pyproject.toml`

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
This PR aims to minimize the number of configuration files for the project. Instead of a separate `.pylintrc` and `setup.cfg`, we now would have all the configurations within the `pyproject.toml` file leading to a overall minimal code structure.

### CODE REVIEW

**Comment by SauravMaheshkar** [comment]
In [27963b1](https://github.com/geoopt/geoopt/pull/201/commits/27963b136adfead91b344f15aa5e57f44af53f58), I added the pylint configuration to `pyproject.toml` using the output generated after running:

```bash
pylint --generate-toml-config
```

**Comment by ferrine** [comment]
Hi! Thank you for looking into this! I'll always put a green light to new contributions!

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
...-ex-returned", "undefined-variable", "invalid-all-format", "no-name-in-module", "potential-index-error", "relative-beyond-top-level", "bad-super-call", "bad-except-order", "raising-bad-type", "misplaced-bare-raise", "bad-exception-cause", "raising-non-exception", "notimplemented-raised", "catching-non-exception", "invalid-envvar-value", "yield-inside-async-function", "not-async-context-manager", "bad-format-character", "truncated-format-string", "mixed-format-string", "format-needs-mapping", "missing-format-string-key", "too-many-format-args", "too-few-format-args", "bad-string-format-type", "bad-str-strip-call", "init-is-generator", "return-in-init", "function-redefined", "not-in-loop", "return-outside-function", "yield-outside-function", "return-arg-in-generator", "nonexistent-operator", "duplicate-argument-name", "abstract-class-instantiated", "too-many-star-expressions", "invalid-star-assignment-target", "star-needs-assignment-target", "nonlocal-and-global", "continue-in-finally", "nonlocal-without-binding", "used-prior-global-declaration", "bad-reversed-sequence", "misplaced-format-function", "fatal", "astroid-error", "parse-error", "config-parse-error", "method-check-failed"]
```
Comment (ferrine) [inline pyproject.toml:372]: This looks to verbose. The options that work generally fine across many of projects I'm involved in are here
https://github.com/pymc-devs/pymc/blob/main/.pylintrc
The idea is to blacklist all and enable some of the checks, it should look more concise

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+
+# Argument names that match this expression will be ignored.
+ignored-argument-names = "_.*"
+
+# Tells whether we should check for unused import in __init__ files.
+# init-import =
+
+# List of qualified module names which can have objects that can redefine
+# builtins.
+redefining-builtins-modules = ["six.moves", "past.builtins", "future.builtins", "builtins", "io"]
```
Comment (ferrine) [inline pyproject.toml:549]: new line

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,549 @@
+[tool.pydocstyle]
```
Comment (ferrine) [inline pyproject.toml:1]: You can also move setup.py configuration inside. This is not involved. Here is the example https://github.com/pymc-devs/pytensor/blob/main/pyproject.toml

**Review by ferrine — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (SauravMaheshkar) [reply pyproject.toml]: Addressed in [254124c](https://github.com/geoopt/geoopt/pull/201/commits/254124c325aa2e672e86327804c077bb4e7f8286)

File: pyproject.toml (hunk 2 above)
Comment (SauravMaheshkar) [reply pyproject.toml]: Addressed in [254124c](https://github.com/geoopt/geoopt/pull/201/commits/254124c325aa2e672e86327804c077bb4e7f8286)

File: pyproject.toml (hunk 3 above)
Comment (SauravMaheshkar) [reply pyproject.toml:40]: Addressed in [b0f2e8b](https://github.com/geoopt/geoopt/pull/201/commits/b0f2e8b8c04fcc0505d629530da370a03c370591)

**Comment by SauravMaheshkar** [comment]
Request for Review @ferrine

**Comment by ferrine** [comment]
Hi, it looks good to me, I’ve approved tests to check everything is ok.

**Comment by SauravMaheshkar** [comment]
[b0f2e8b](https://github.com/geoopt/geoopt/pull/201/commits/b0f2e8b8c04fcc0505d629530da370a03c370591), failed because I left out some empty arguments. Addressed in [b344d52](https://github.com/geoopt/geoopt/pull/201/commits/b344d52606084e54288a09be866890d1f1974a81)

**Comment by SauravMaheshkar** [comment]
Added the explicit use of `pyproject.toml` in [93a26c8](https://github.com/geoopt/geoopt/pull/201/commits/93a26c886c4b516937f9e030b179b48f1dda84e9)

**Comment by SauravMaheshkar** [comment]
`pydocstyle` didn't adopt `pyproject.toml` until a new version, that's why the version constraint in [6ceaa1c](https://github.com/geoopt/geoopt/pull/201/commits/6ceaa1c03b7915837280c1f145b6cc2fab91ced5)

**Comment by ferrine** [comment]
There is an issue with versioneer. I've tried to install the PR branch and I get incorrect version

```console
pip install -e .
Obtaining file:///Users/ferres/Documents/Github/geoopt
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Installing backend dependencies ... done
  Preparing editable metadata (pyproject.toml) ... done
Requirement already satisfied: numpy in /Users/ferres/micromamba/envs/geoopt/lib/python3.9/site-packages (from geoopt==0.0.0) (1.24.2)
Requirement already satisfied: scipy in /Users/ferres/micromamba/envs/geoopt/lib/python3.9/site-packages (from geoopt==0.0.0) (1.10.0)
Requirement already satisfied: torch!=1.12.*,>=1.11.0 in /Users/ferres/micromamba/envs/geoopt/lib/python3.9/site-packages (from geoopt==0.0.0) (1.13.1)
Requirement already satisfied: typing-extensions in /Users/ferres/micromamba/envs/geoopt/lib/python3.9/site-packages (from torch!=1.12.*,>=1.11.0->geoopt==0.0.0) (4.4.0)
Building wheels for collected packages: geoopt
  Building editable for geoopt (pyproject.toml) ... done
  Created wheel for geoopt: filename=geoopt-0.0.0-0.editable-py3-none-any.whl size=13194 sha256=aa4262cabbd8bb6b6175ee7076aae3cb4a1a176eaa83ee1ea7fd966afc60320e
  Stored in directory: /private/var/folders/rx/rk9gm4ln35z802s3p81wfz8r0000gp/T/pip-ephem-wheel-cache-dt7m8rzm/wheels/b9/30/27/13ba375ef822651ec0db8aafa8f841d996bfdff903cbcf0adc
Successfully built geoopt
Installing collected packages: geoopt
  Attempting uninstall: geoopt
    Found existing installation: geoopt 0.5.0
    Uninstalling geoopt-0.5.0:
      Successfully uninstalled geoopt-0.5.0
Successfully installed geoopt-0.0.0
```

**Comment by SauravMaheshkar** [comment]
> There is an issue with versioneer. I've tried to install the PR branch and I get incorrect version
> 
> ```
> pip install -e .
> Obtaining file:///Users/ferres/Documents/Github/geoopt
>   Installing build dependencies ... done
>   Checking if build backend supports build_editable ... done
>   Getting requirements to build editable ... done
>   Installing backend dependencies ... done
>   Preparing editable metadata (pyproject.toml) ... done
> Requirement already satisfied: numpy in /Users/ferres/micromamba/envs/geoopt/lib/python3.9/site-packages (from geoopt==0.0.0) (1.24.2)
> Requirement already satisfied: scipy in /Users/ferres/micromamba/envs/geoopt/lib/python3.9/site-packages (from geoopt==0.0.0) (1.10.0)
> Requirement already satisfied: torch!=1.12.*,>=1.11.0 in /Users/ferres/micromamba/envs/geoopt/lib/python3.9/site-packages (from geoopt==0.0.0) (1.13.1)
> Requirement already satisfied: typing-extensions in /Users/ferres/micromamba/envs/geoopt/lib/python3.9/site-packages (from torch!=1.12.*,>=1.11.0->geoopt==0.0.0) (4.4.0)
> Building wheels for collected packages: geoopt
>   Building editable for geoopt (pyproject.toml) ... done
>   Created wheel for geoopt: filename=geoopt-0.0.0-0.editable-py3-none-any.whl size=13194 sha256=aa4262cabbd8bb6b6175ee7076aae3cb4a1a176eaa83ee1ea7fd966afc60320e
>   Stored in directory: /private/var/folders/rx/rk9gm4ln35z802s3p81wfz8r0000gp/T/pip-ephem-wheel-cache-dt7m8rzm/wheels/b9/30/27/13ba375ef822651ec0db8aafa8f841d996bfdff903cbcf0adc
> Successfully built geoopt
> Installing collected packages: geoopt
>   Attempting uninstall: geoopt
>     Found existing installation: geoopt 0.5.0
>     Uninstalling geoopt-0.5.0:
>       Successfully uninstalled geoopt-0.5.0
> Successfully installed geoopt-0.0.0
> ```
> 
> These instructions should help to figure out the issue
> 
> https://github.com/python-versioneer/python-versioneer/blob/master/INSTALL.md

Doesn't `versioneer` require a `setup.py` file ? This PR aims to remove the `setup.py` and use only a `pyproject.toml`. 

If automatic versioning is desired then we can just use the `flit` backend to automatically fetch the `__version__` attribute. For reference, when I ran `pip install -e .` with the modified `pyproject.toml` with a `flit` backend.

```lang-sh
Obtaining file:///Users/saurav/github/geoopt
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Requirement already satisfied: torch!=1.12.*,>=1.11.0 in /opt/homebrew/lib/python3.8/site-packages (from geoopt==0.5.0) (1.13.1)
Requirement already satisfied: scipy in /opt/homebrew/lib/python3.8/site-packages (from geoopt==0.5.0) (1.10.0)
Requirement already satisfied: numpy in /opt/homebrew/lib/python3.8/site-packages (from geoopt==0.5.0) (1.24.2)
Requirement already satisfied: typing-extensions in /opt/homebrew/lib/python3.8/site-packages (from torch!=1.12.*,>=1.11.0->geoopt==0.5.0) (4.4.0)
Building wheels for collected packages: geoopt
  Building editable for geoopt (pyproject.toml) ... done
  Created wheel for geoopt: filename=geoopt-0.5.0-py3-none-any.whl size=7839 sha256=6075d368d7f9728d64d1d85e7ddd774d8fcc1411d840c61752b5351b87574574
  Stored in directory: /private/var/folders/z3/0svgl03x155444l4v7pdlg_40000gn/T/pip-ephem-wheel-cache-wtxgrx6m/wheels/f3/d1/01/b041546afc1d9c90980103872e30c2d19c939e8d492d1a4cf6
Successfully built geoopt
Installing collected packages: geoopt
  Attempting uninstall: geoopt
    Found existing installation: geoopt 0.5.0
    Uninstalling geoopt-0.5.0:
      Successfully uninstalled geoopt-0.5.0
Successfully installed geoopt-0.5.0
```

**Comment by ferrine** [comment]
Yes, let's switch

**Comment by SauravMaheshkar** [comment]
Done in [7bb9343](https://github.com/geoopt/geoopt/pull/201/commits/7bb9343e14808f08f9fb7e5f4a583631653f4359)

**Comment by ferrine** [comment]
thank you!

**Pull request merged**